### PR TITLE
[REG-1463] Generate 'partial' class to attach RGState and RGAction to host behavior

### PR DESCRIPTION
**Pull request by analogrelay**
This updates our Unity code gen to add partial classes that create a `RequireComponent` dependency from the "host" behavior (the `MonoBehavior` on which the `[RGState]` or `[RGAction]` attribute is placed) to the generated `RGState_[HostBehaviorName]` and `RGAction_[HostBehaviorName]` components. This signals to the Unity editor, when reimporting assets, to add the `RGState_...`/`RGAction_...` components to any `GameObject` that also has the host behavior attached. This mostly allows us to "auto-attach" the behaviors.

From my testing, this is fairly consistent. Unity [does caution that it doesn't check dependencies in all scenarios](https://docs.unity3d.com/ScriptReference/RequireComponent.html), but it does it whenever `GameObject.AddComponent` is called.

In order to generate this auto-attach class, we need the original host behavior to be marked as a `partial` class. I added code to check for that and report an error if that's not the case. WIth the changes in #105, this also causes a dialog to pop up to draw your attention to the log:

image

NOTE: One additional change I made is to restore the "Generate Scripts" menu item. After #105 merged, auto-generation is mostly working but there are a few scenarios where I've noticed it doesn't work:

1. If you change anything in the RegressionGames packages, it doesn't re-generate. This _mostly_ affects us
2. If you delete or otherwise mess with the generated code (perhaps accidentally).

Both cases are fairly rare (case 1 is mostly just for us), but it seems reasonable to me to bring back the _ability_ to manually regenerate the scripts, even if it's only needed in edge cases.

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: Editor/Scripts/CodeGenerators/GenerateRGStateClasses.cs [hunk 1]
```diff
@@ -97,12 +103,13 @@ public static void Generate(List<RGStateAttributesInfo> rgStateAttributesInfos)
                                     GenerateStateEntityFields(rgStateAttributeInfo.State)
                                 ),
                             classDeclaration
-
                         );
 
                     // Add the namespace declaration to the compilation unit
                     compilationUnit = compilationUnit.AddMembers(namespaceDeclaration);
 
+                    compilationUnit = compilationUnit.AddMembers(CodeGeneratorUtils.CreatePartialAttacherClass(rgStateAttributeInfo.NameSpace, componentType, $"RegressionGames.RGBotConfigs.{className}"));
```
Comment (RG-nAmKcAz) [inline Editor/Scripts/CodeGenerators/GenerateRGStateClasses.cs:111]: I think we should consider this having a `RequireComponent` for the behavior we found the RGState or RGAction attribute in as well.  Thus if they remove that behavior they will get warned that our scripts require it.

**Review by RG-nAmKcAz — CHANGES_REQUESTED** [review]

**Review by rcornwall-reg — APPROVED** [review]

**Review by RG-nAmKcAz — APPROVED** [review]

**Comment by analogrelay** [comment]
Resolved a gnarly enough conflict that I'd like to get re-reviews (particularly from @RG-nAmKcAz since that's where the conflict arose and I want to make sure I didn't blast away his changes). I'll do another smoke-test to make sure things are working.

And I see you already approved 🤣 . Thanks @RG-nAmKcAz !

**Comment by RG-nAmKcAz** [comment]
> Resolved a gnarly enough conflict that I'd like to get re-reviews (particularly from @RG-nAmKcAz since that's where the conflict arose and I want to make sure I didn't blast away his changes). I'll do another smoke-test to make sure things are working.
> 
> And I see you already approved 🤣 . Thanks @RG-nAmKcAz !

For clarity, I'm not hand inspecting the code to see if you merged it correctly.  I would expect the dev themselves to do that as necessary.  I didn't see anything obvious though at a high level glance.

**Comment by analogrelay** [comment]
> For clarity, I'm not hand inspecting the code to see if you merged it correctly. I would expect the dev themselves to do that as necessary. I didn't see anything obvious though at a high level glance.

Yep, that's my expectation too, but I think it's always valuable for the conflicting commit author to take a scan over it. Thanks for looking!

**Comment by analogrelay** [comment]
There was one small style change lost in the conflict resolution. I've restored that: https://github.com/Regression-Games/RGUnityBots/pull/116/commits/16e8995f513e22ab571db7e72eb0654656565f93

**Pull request merged**